### ISSUE TITLE
remove features from textstat_collocations

### ISSUE THREAD

**Issue by danilovcorrea** [quanteda/quanteda#1754]
Dear,

After creating a dictionary through a lot of steps (characters -> tokens -> dfm  -> dfm_trim -> dfm_weight), i have discovered the amazing function **textstat_collocations** that synthesizes what is needed to build the text statistics directly through characters, saving a lot a steps.

But i'm in doubt:

- The arguments description states that _"the tokens object should include punctuation, and if any words have been removed, these should have been removed with padding = TRUE. While identifying collocations for tokens objects is supported, you will get better results with character or corpus objects due to relatively imperfect detection of sentence boundaries from texts already tokenized."_

I need to remove sweary words (like ones available at sweary package [https://github.com/pdrhlik/sweary](https://github.com/pdrhlik/sweary)), punctuation, numbers, separators, twitter, hyphens and url. It's possible to remove those during or after the creation of the textstat_collocations ?

**Comment by kbenoit** [comment]
`textstat_collocations()` detects multi-word expressions based on frequency and association of collocated tokens. The reason it's recommended to leave the pads is that otherwise, `textstat_collocations()` will consider as adjacent tokens that were in fact separated by the punctuation etc that has been removed.

Because of the way that removals are processed in `tokens()`, there is no option to remove pads there. Better to tokenize all and then remove what you want using patterns. Below, the `\\p{P}` is for punctuation and the `\\p{N}` is for numbers.

``` r
library("quanteda")
#> Package version: 1.5.3
txt <- "I don't give 2 damns, whether you say yes, or no. You say yes, I say no."

toks <- tokens(txt) %>%
    tokens_remove(pattern = "^[\\p{P}\\p{N}]+$", valuetype = "regex",
                  padding = TRUE)
textstat_collocations(toks)
#>   collocation count count_nested length   lambda        z
#> 1     you say     2            0      2 4.121744 2.333208
#> 2     say yes     2            0      2 4.121744 2.333208
```

<sup>Created on 2019-10-15 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by danilovcorrea** [comment]
Hi @kbenoit , thanks in advance,

But it doesn't remove numbers or punctuation inside tokens, like:
```r
txt <- "8-PM #wf 8-PM #wf 8-PM @mail 8-PM 12th #wf #wf 12th > \U0001f33a > \U0001f33a > http:// @RT @mail 8-PM"
toks <- tokens(txt) %>%
    tokens_remove(pattern = "^[\\p{P}\\p{N}]+$", valuetype = "regex",
                  padding = TRUE)
textstat_collocations(toks)
#>      collocation count count_nested length   lambda        z
#> 1 <U+0001F33A> >     2            0      2 4.174387 2.364059
#> 2 > <U+0001F33A>     2            0      2 4.174387 2.364059
#> 3     @mail 8-pm     2            0      2 3.610918 2.137404
#> 4       8-pm #wf     2            0      2 1.887070 1.680731
#> 5       #wf 8-pm     2            0      2 1.887070 1.680731
```
One incredible vantage of tokens() is the possibility to pass the arguments  remove_numbers = T,                        remove_punct = T, remove_separators = T, remove_twitter = T, remove_hyphens = T, remove_url = T.

It saves a lot of work trying to clean tokens removing regex patterns. How can i use this advantage to produce the textstat_collocations?

In other words, what's is the fastest way to create a clean (only words) textstat_collocations from characters?

**Comment by kbenoit** [comment]
You can use 
```r
tokens_remove(x, pattern = "\\d", valuetype = "regex", pad = TRUE)
```
 for instance to remove any token containing a numeral. All of the `remove_*` options in `tokens()` can be recreated using `tokens_remove()`, if you know how to create the appropriate regular expression.
